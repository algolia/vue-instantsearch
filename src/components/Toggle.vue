<template>
  <label :class="isRefined ? bem('label', 'active') : bem('label')">
    <input type="checkbox" :class="bem('checkbox')" v-model="isRefined"> 
    <slot :label="label" :count="count" :active="isRefined">
      <span :class="bem('label')">{{ label }}</span>
      <span :class="bem('count')">{{ count }}</span>
    </slot>
  </label>
</template>

<script>import { FACET_OR } from '../store';
import algoliaComponent from '../component';

export default {
  mixins: [algoliaComponent],
  props: {
    attributeName: {
      type: String,
      required: true,
    },
    value: {
      default: true,
    },
    label: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      blockClassName: 'ais-toggle',
    };
  },
  created() {
    this.searchStore.addFacet(this.attributeName, FACET_OR);
  },
  destroyed() {
    this.searchStore.removeFacet(this.attributeName);
  },
  computed: {
    facetValue() {
      return this.searchStore.getFacetValue(this.attributeName, this.value);
    },
    isRefined: {
      set(value) {
        if (value === true) {
          this.enable();
        } else {
          this.disable();
        }
      },
      get() {
        return this.facetValue.isRefined;
      },
    },
    count() {
      return this.facetValue.count;
    },
  },
  methods: {
    enable() {
      this.searchStore.addFacetRefinement(this.attributeName, this.value);
    },
    disable() {
      this.searchStore.removeFacetRefinement(this.attributeName, this.value);
    },
  },
};
</script>
