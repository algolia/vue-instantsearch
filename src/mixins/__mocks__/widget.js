let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
