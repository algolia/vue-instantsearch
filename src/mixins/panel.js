import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
