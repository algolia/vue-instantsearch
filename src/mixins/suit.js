import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
