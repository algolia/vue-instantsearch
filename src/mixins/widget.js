const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
