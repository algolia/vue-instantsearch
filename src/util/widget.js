export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
