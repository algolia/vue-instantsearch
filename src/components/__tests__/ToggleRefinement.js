import { mount } from '@vue/test-utils';
import { __setState } from '../../mixins/widget';
import Toggle from '../ToggleRefinement.vue';

jest.mock('../../mixins/widget');
jest.mock('../../mixins/panel');

const defaultValue = {
  name: 'free_shipping',
  count: 100,
  isRefined: false,
};

const defaultState = {
  value: defaultValue,
  refine: () => {},
  createURL: () => {},
};

const defaultProps = {
  attribute: 'free_shipping',
  label: 'Free Shipping',
};

it('accepts an attribute prop', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Toggle, {
    propsData: defaultProps,
  });

  expect(wrapper.vm.widgetParams.attribute).toBe('free_shipping');
});

it('accepts a label prop', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Toggle, {
    propsData: defaultProps,
  });

  expect(wrapper.vm.widgetParams.label).toBe('Free Shipping');
});

it('accepts an on prop', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Toggle, {
    propsData: {
      ...defaultProps,
      on: 'somevalue',
    },
  });

  expect(wrapper.vm.widgetParams.on).toBe('somevalue');
});

it('accepts an off prop', () => {
  __setState({ ...defaultState });

  const wrapper = mount(Toggle, {
    propsData: {
      ...defaultProps,
      off: false,
    },
  });

  expect(wrapper.vm.widgetParams.off).toBe(false);
});

describe('default render', () => {
  it('renders correctly', () => {
    __setState({ ...defaultState });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly without refinement (with 0)', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        count: 0,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly without refinement (with null)', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        count: null,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly with the value not selected', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        isRefined: false,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    expect(wrapper.find('input').element.checked).toBe(false);
  });

  it('renders correctly with the value selected', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        isRefined: true,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    expect(wrapper.find('input').element.checked).toBe(true);
  });

  it('calls refine on input change', () => {
    const refine = jest.fn();

    __setState({
      ...defaultState,
      refine,
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
    });

    wrapper.find('input').setChecked(true);

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultValue);
  });
});

it('calls the Panel mixin with `value.count`', () => {
  __setState({
    ...defaultState,
    value: {
      // Otherwise setData update the default value
      // and impact the other tests. We should not
      // rely on a global state for the tests.
      ...defaultValue,
    },
  });

  const wrapper = mount(Toggle, {
    propsData: defaultProps,
  });

  const mapStateToCanRefine = () =>
    wrapper.vm.mapStateToCanRefine(wrapper.vm.state);

  expect(mapStateToCanRefine()).toBe(true);

  wrapper.setData({
    state: {
      value: {
        count: 0,
      },
    },
  });

  expect(mapStateToCanRefine()).toBe(false);

  expect(wrapper.vm.mapStateToCanRefine({})).toBe(false);
});

describe('custom default render', () => {
  const defaultScopedSlot = `
    <a
      slot-scope="{ value, canRefine, refine, createURL }"
      :href="createURL()"
      :class="[!canRefine && 'noRefinement']"
      @click.prevent="refine(value)"
    >
      <span>{{ value.name }}</span>
      <span>{{ value.isRefined ? '(is enabled)' : '(is disabled)' }}</span>
    </a>
  `;

  it('renders correctly', () => {
    __setState({ ...defaultState });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly without refinement', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        count: 0,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly with a URL for the href', () => {
    __setState({
      ...defaultState,
      createURL: () => `/free-shipping`,
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly with the value selected', () => {
    __setState({
      ...defaultState,
      value: {
        ...defaultValue,
        isRefined: true,
      },
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('calls refine on link click', () => {
    const refine = jest.fn();

    __setState({
      ...defaultState,
      refine,
    });

    const wrapper = mount(Toggle, {
      propsData: defaultProps,
      scopedSlots: {
        default: defaultScopedSlot,
      },
    });

    wrapper.find('a').trigger('click');

    expect(refine).toHaveBeenCalledTimes(1);
    expect(refine).toHaveBeenCalledWith(defaultValue);
  });
});
