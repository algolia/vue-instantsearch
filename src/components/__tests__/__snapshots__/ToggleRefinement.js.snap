// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom default render renders correctly 1`] = `

<div class="ais-ToggleRefinement">
  <a class>
    <span>
      free_shipping
    </span>
    <span>
      (is disabled)
    </span>
  </a>
</div>

`;

exports[`custom default render renders correctly with a URL for the href 1`] = `

<div class="ais-ToggleRefinement">
  <a href="/free-shipping"
     class
  >
    <span>
      free_shipping
    </span>
    <span>
      (is disabled)
    </span>
  </a>
</div>

`;

exports[`custom default render renders correctly with the value selected 1`] = `

<div class="ais-ToggleRefinement">
  <a class>
    <span>
      free_shipping
    </span>
    <span>
      (is enabled)
    </span>
  </a>
</div>

`;

exports[`custom default render renders correctly without refinement 1`] = `

<div class="ais-ToggleRefinement ais-ToggleRefinement--noRefinement">
  <a class="noRefinement">
    <span>
      free_shipping
    </span>
    <span>
      (is disabled)
    </span>
  </a>
</div>

`;

exports[`default render renders correctly 1`] = `

<div class="ais-ToggleRefinement">
  <label class="ais-ToggleRefinement-label">
    <input type="checkbox"
           name="free_shipping"
           class="ais-ToggleRefinement-checkbox"
           value="true"
    >
    <span class="ais-ToggleRefinement-labelText">
      Free Shipping
    </span>
    <span class="ais-ToggleRefinement-count">
      100
    </span>
  </label>
</div>

`;

exports[`default render renders correctly without refinement (with 0) 1`] = `

<div class="ais-ToggleRefinement ais-ToggleRefinement--noRefinement">
  <label class="ais-ToggleRefinement-label">
    <input type="checkbox"
           name="free_shipping"
           class="ais-ToggleRefinement-checkbox"
           value="true"
    >
    <span class="ais-ToggleRefinement-labelText">
      Free Shipping
    </span>
    <span class="ais-ToggleRefinement-count">
      0
    </span>
  </label>
</div>

`;

exports[`default render renders correctly without refinement (with null) 1`] = `

<div class="ais-ToggleRefinement ais-ToggleRefinement--noRefinement">
  <label class="ais-ToggleRefinement-label">
    <input type="checkbox"
           name="free_shipping"
           class="ais-ToggleRefinement-checkbox"
           value="true"
    >
    <span class="ais-ToggleRefinement-labelText">
      Free Shipping
    </span>
  </label>
</div>

`;
